### PR TITLE
Clean list of Operations supported by fence and quiet

### PR DESCRIPTION
**Pull request by naveen-rn**
Fix for #278

Duplicate of https://github.com/openshmem-org/specification/pull/356 - with the current section committee branch

### CODE REVIEW

**Comment by naveen-rn** [comment]
@nspark and @bryantlam please review the changes to get it merged into the section committee branch.

File: main_spec.tex [hunk 1]
```diff
@@ -473,7 +473,28 @@ \subsection{Memory Ordering Routines}\label{subsec:memory_order}
 ensure ordering and/or delivery of memory store, blocking \OPR{Put}, \ac{AMO},
 and \OPR{put-with-signal}, as well as nonblocking \PUT{},
 \OPR{put-with-signal}, \GET{}, and \ac{AMO} routines to symmetric data
-objects.
+objects. Table~\ref{mem-order} refers to the list of operations that is ensured
+ordering and guaranteed completion by memory ordering routines.
+
+\begin{table}[h]
+  \begin{center}
+    \begin{tabular}{|l|l|l|}
```
Comment (nspark) [inline main_spec.tex]:
```suggestion
    \begin{tabular}{|l|c|c|}
```

File: main_spec.tex [hunk 2]
```diff
@@ -473,7 +473,28 @@ \subsection{Memory Ordering Routines}\label{subsec:memory_order}
 ensure ordering and/or delivery of memory store, blocking \OPR{Put}, \ac{AMO},
 and \OPR{put-with-signal}, as well as nonblocking \PUT{},
 \OPR{put-with-signal}, \GET{}, and \ac{AMO} routines to symmetric data
-objects.
+objects. Table~\ref{mem-order} refers to the list of operations that is ensured
+ordering and guaranteed completion by memory ordering routines.
```
Comment (nspark) [inline main_spec.tex]:
```suggestion
objects. Table~\ref{mem-order} lists the operations for which \openshmem
memory ordering routines provide guarantees of ordering or completion.
```

File: content/shmem_fence.tex [hunk 3]
```diff
@@ -19,18 +16,14 @@
...
-    \OPR{put-with-signal}, and \ac{AMO}
-    routines to symmetric data objects.  All memory store, blocking \PUT{},
-    \ac{AMO}, and \OPR{put-with-signal}, as well as nonblocking \PUT{},
-    \OPR{put-with-signal}, and \ac{AMO}
-    routines to symmetric data objects issued to a particular remote \ac{PE}
-    on the given context prior
+    This routine assures ordering of delivery of operations to symmetric data
+    objects.
+    Refer Table~\ref{mem-order} for the list of operations that are ordered
+    by \FUNC{shmem\_fence} routine.
```
Comment (nspark) [inline content/shmem_fence.tex]:
```suggestion
    Table~\ref{mem-order} lists the operations that are ordered
    by the \FUNC{shmem\_fence} routine.
```

File: content/shmem_fence.tex [hunk 4]
```diff
@@ -1,8 +1,5 @@
 \apisummary{
-    Assures ordering of delivery of memory store, blocking \PUT{}, 
-    \ac{AMO}, and \OPR{put-with-signal}, as well as nonblocking 
-    \PUT{}, \OPR{put-with-signal}, and \ac{AMO}
-    routines to symmetric data objects.
+    Assures ordering of delivery of operations to symmetric data objects.
```
Comment (nspark) [inline content/shmem_fence.tex]:
```suggestion
    Assures ordering of delivery of operations on symmetric data objects.
```

File: content/shmem_fence.tex [hunk 5]
```diff
@@ -19,18 +16,14 @@
...
 \apidescription{
-    This routine assures ordering of delivery of memory store, blocking \PUT{},
-    \ac{AMO}, and \OPR{put-with-signal}, as well as nonblocking \PUT{},
-    \OPR{put-with-signal}, and \ac{AMO}
-    routines to symmetric data objects.  All memory store, blocking \PUT{},
-    \ac{AMO}, and \OPR{put-with-signal}, as well as nonblocking \PUT{},
-    \OPR{put-with-signal}, and \ac{AMO}
-    routines to symmetric data objects issued to a particular remote \ac{PE}
-    on the given context prior
+    This routine assures ordering of delivery of operations to symmetric data
```
Comment (nspark) [inline content/shmem_fence.tex]:
```suggestion
    This routine assures ordering of delivery of operations on symmetric data
```

File: content/shmem_fence.tex [hunk 6]
```diff
@@ -19,18 +16,14 @@
...
+    objects.
+    Refer Table~\ref{mem-order} for the list of operations that are ordered
+    by \FUNC{shmem\_fence} routine.
+    All operations to symmetric data objects issued to a particular
```
Comment (nspark) [inline content/shmem_fence.tex]:
```suggestion
    All operations on symmetric data objects issued to a particular
```

File: content/shmem_fence.tex [hunk 7]
```diff
@@ -19,18 +16,14 @@
...
+    remote \ac{PE} on the given context prior
```
Comment (nspark) [inline content/shmem_fence.tex]: Do we really want to keep "remote" here, since `shmem_fence` also orders operations operating on the calling PE?

File: content/shmem_fence.tex [hunk 8]
```diff
@@ -19,18 +16,14 @@
...
     to the call to \FUNC{shmem\_fence} are guaranteed to be delivered before any
-    subsequent memory store, blocking \PUT{}, \ac{AMO}, and \OPR{put-with-signal}, 
-    as well as nonblocking \PUT{}, \OPR{put-with-signal}, and \ac{AMO}
-    routines to symmetric data
+    subsequent operations to symmetric data
```
Comment (nspark) [inline content/shmem_fence.tex]:
```suggestion
    subsequent operations on symmetric data
```

File: content/shmem_fence.tex [hunk 9]
```diff
@@ -51,14 +44,12 @@
     ordering of its memory accesses.
     There is a subtle difference between
     \FUNC{shmem\_fence} and \FUNC{shmem\_quiet}, in that, \FUNC{shmem\_quiet}
-    guarantees completion of memory store, blocking \PUT{}, \ac{AMO}, and
-    \OPR{put-with-signal}, as well as nonblocking \PUT{}, \OPR{put-with-signal}, and \ac{AMO} routines to
+    guarantees completion of all operations to
```
Comment (nspark) [inline content/shmem_fence.tex]:
```suggestion
    guarantees completion of all operations on
```

File: content/shmem_fence.tex [hunk 10]
```diff
@@ -51,14 +44,12 @@
...
     symmetric data objects which makes the updates visible to all other
     \acp{PE}.
 
-    The \FUNC{shmem\_quiet} routine should be called if completion of memory
-    store, blocking \PUT{}, \ac{AMO}, and \OPR{put-with-signal}, as well as
-    nonblocking \PUT{}, \OPR{put-with-signal}, and \ac{AMO} routines to symmetric data objects is desired
+    The \FUNC{shmem\_quiet} routine should be called if completion of operations
+    to symmetric data objects is desired
```
Comment (nspark) [inline content/shmem_fence.tex]:
```suggestion
    on symmetric data objects is desired
```

File: content/shmem_fence.tex [hunk 11]
```diff
@@ -51,14 +44,12 @@
...
     when multiple remote \acp{PE} are involved.
```
Comment (nspark) [inline content/shmem_fence.tex]: Same about "remote" as above.

File: content/shmem_quiet.tex [hunk 12]
```diff
@@ -1,8 +1,6 @@
 \apisummary{
-    Waits for completion of all outstanding memory store, blocking 
-    \PUT{}, \ac{AMO}, and \OPR{put-with-signal}, as well as 
-    nonblocking \PUT{}, \OPR{put-with-signal}, \GET{}, and \ac{AMO} routines 
-    to symmetric data objects issued by a \ac{PE}.
+    Waits for completion of all outstanding operations on symmetric data objects
```
Comment (nspark) [inline content/shmem_quiet.tex]:
```suggestion
    Waits for completion of outstanding operations on symmetric data objects
```

Suggest removing "all", as the outstanding operations are limited in scope to a (implicit or explicit) context.

File: content/shmem_quiet.tex [hunk 13]
```diff
@@ -19,14 +17,12 @@
...
-    \OPR{put-with-signal}, as well as nonblocking \PUT{},\OPR{put-with-signal}, \GET{}, and \ac{AMO} routines on
-    symmetric data objects issued by the calling \ac{PE} on the given context.
-    All memory store, blocking \PUT{}, \ac{AMO}, and \OPR{put-with-signal}, as
-    well as nonblocking \PUT{}, \OPR{put-with-signal}, \GET{}, and \ac{AMO} routines to
-    symmetric data objects are guaranteed to be completed and visible to all
-    \acp{PE} when \FUNC{shmem\_quiet} returns.
+    The \FUNC{shmem\_quiet} routine ensures completion of all operations
+    on symmetric data objects issued by the calling \ac{PE} on the given context.
+    Refer Table~\ref{mem-order} for the list of operations that are ensured
+    completion by \FUNC{shmem\_quiet} routine.
```
Comment (nspark) [inline content/shmem_quiet.tex]:
```suggestion
    Table~\ref{mem-order} lists the operations for which the
    \FUNC{shmem\_quiet} routine ensures completion..
```

File: content/shmem_quiet.tex [hunk 14]
```diff
@@ -19,14 +17,12 @@
...
+    All operations to symmetric data objects are guaranteed to be completed and
```
Comment (nspark) [inline content/shmem_quiet.tex]:
```suggestion
    All operations on symmetric data objects are guaranteed to be complete and
```

File: content/shmem_quiet.tex [hunk 15]
```diff
@@ -38,23 +34,21 @@
 
 \apinotes{
     \FUNC{shmem\_quiet} is most useful as a way of ensuring completion of
-    several memory store, blocking \PUT{}, \ac{AMO}, and \OPR{put-with-signal},
-    as well as nonblocking \PUT{},
-    \OPR{put-with-signal}, \GET{}, and \ac{AMO} routines to symmetric data objects initiated by the calling
-    \ac{PE}.  For example, one might use \FUNC{shmem\_quiet} to await delivery
+    several operations to symmetric data objects initiated by the calling
```
Comment (nspark) [inline content/shmem_quiet.tex]:
```suggestion
    several operations on symmetric data objects initiated by the calling
```

**Review by nspark — DISMISSED** [review]
- Suggest rewording for clarity and readability
- Change "to" to "on" for consistency

File: content/shmem_fence.tex (hunk 7 above)
Comment (naveen-rn) [reply content/shmem_fence.tex]: I think we should remove ```remote``` from here. 

Also, in general, do we need to say something about the operation of the ```shmem_ctx_fence``` with respect to ```SHMEM_CTX_NOSTORE```. Reading the NOSTORE, it kind of confuses how it related to the stores in this description.

**Comment by naveen-rn** [comment]
@nspark regarding the "remote" usage - I think we should remove remote from here.

Also, in general, do we need to say something about the operation of the shmem_ctx_fence with respect to SHMEM_CTX_NOSTORE. Reading the NOSTORE, it kind of confuses how it related to the stores in this description.

@bryantlam can you provide some input on this PR?

File: content/shmem_fence.tex (hunk 7 above)
Comment (BryantLam) [reply content/shmem_fence.tex]: I'm neutral on whether to handle `SHMEM_CTX_NOSTORE` here. The description of the NOSTORE option in `shmem_ctx_create` creates an exception to the behaviors listed here in `shmem_fence`. That seems sufficient to me, but I'd also be okay if something was added here since the potential for overly duplicating text is limited to fence and quiet.

File: content/shmem_fence.tex [hunk 16]
```diff
@@ -51,14 +44,12 @@
...
     \acp{PE}.
 
-    The \FUNC{shmem\_quiet} routine should be called if completion of memory
-    store, blocking \PUT{}, \ac{AMO}, and \OPR{put-with-signal}, as well as
-    nonblocking \PUT{}, \OPR{put-with-signal}, and \ac{AMO} routines to symmetric data objects is desired
+    The \FUNC{shmem\_quiet} routine should be called if completion of operations
+    on symmetric data objects is desired
     when multiple remote \acp{PE} are involved.
 
     In an \openshmem program with multithreaded \acp{PE}, it is the
```
Comment (BryantLam) [inline content/shmem_fence.tex:55]: I think you missed one: `(e.g. \PUT{}, ... and nonblocking routines)` in this paragraph can be removed too.

File: content/shmem_quiet.tex [hunk 17]
```diff
@@ -19,14 +17,12 @@
...
-    \OPR{put-with-signal}, as well as nonblocking \PUT{},\OPR{put-with-signal}, \GET{}, and \ac{AMO} routines on
-    symmetric data objects issued by the calling \ac{PE} on the given context.
-    All memory store, blocking \PUT{}, \ac{AMO}, and \OPR{put-with-signal}, as
-    well as nonblocking \PUT{}, \OPR{put-with-signal}, \GET{}, and \ac{AMO} routines to
-    symmetric data objects are guaranteed to be completed and visible to all
-    \acp{PE} when \FUNC{shmem\_quiet} returns.
+    The \FUNC{shmem\_quiet} routine ensures completion of all operations
+    on symmetric data objects issued by the calling \ac{PE} on the given context.
+    Table~\ref{mem-order} lists the operations for which the
+    \FUNC{shmem\_quiet} routine ensures completion..
```
Comment (BryantLam) [inline content/shmem_quiet.tex]: DocEdit: Remove the double period at the end of this line.

File: content/shmem_quiet.tex [hunk 18]
```diff
@@ -38,23 +34,21 @@
...
+    \ac{PE}. For example, one might use \FUNC{shmem\_quiet} to await delivery
     of a block of data before issuing another \PUT{} or nonblocking
     \PUT{} routine, which sets a completion flag on another \ac{PE}.
-     \FUNC{shmem\_quiet} is not usually needed if
+    \FUNC{shmem\_quiet} is not usually needed if
     \FUNC{shmem\_barrier\_all} or \FUNC{shmem\_barrier} are called.  The barrier
-    routines wait for the completion of outstanding writes (memory store,
-    blocking \PUT{}, \ac{AMO}, and \OPR{put-with-signal}, as well as nonblocking \PUT{},
-    \OPR{put-with-signal}, \GET{}, and \ac{AMO} routines) to
+    routines wait for the completion of outstanding writes to
```
Comment (BryantLam) [inline content/shmem_quiet.tex]: "outstanding writes" might be more consistently phrased as **"outstanding operations"**.  "outstanding requests" would also be okay to me.

File: content/shmem_quiet.tex [hunk 19]
```diff
@@ -38,23 +34,21 @@
...
     symmetric data objects on all \acp{PE}.
 
     In an \openshmem program with multithreaded \acp{PE}, it is the
-    user's responsibility to ensure ordering between operations issued by the threads
-    in a \ac{PE} that target symmetric memory (e.g. \PUT{}, \ac{AMO}, \OPR{put-with-signal}, memory stores,
-    and nonblocking routines) and calls by threads in that \ac{PE} to
+    user's responsibility to ensure ordering between operations issued by the
+    threads in a \ac{PE} that target symmetric memory
+    (e.g. \PUT{}, \ac{AMO}, \OPR{put-with-signal}, memory stores,
+    and nonblocking routines)
```
Comment (BryantLam) [inline content/shmem_quiet.tex]: Similar to my note in `shmem_fence` above, is there a reason this `(e.g. ...)` wasn't omitted?

File: main_spec.tex [hunk 20]
```diff
@@ -473,7 +473,28 @@ \subsection{Memory Ordering Routines}\label{subsec:memory_order}
...
+      Blocking \OPR{Put}                 & X & X \\
+      Blocking \OPR{AMO}                 & X & X \\
+      Blocking \OPR{put-with-signal}     & X & X \\
+      Non-blocking \OPR{Put}             & X & X \\
+      Non-blocking \OPR{AMO}             & X & X \\
+      Non-blocking \OPR{put-with-signal} & X & X \\
+      Non-blocking \OPR{Get}             &   & X \\ \hline
+    \end{tabular}
+    \TableCaptionRef{List of operations ensured ordering and delivery by Memory
+    Ordering routines}
```
Comment (BryantLam) [inline main_spec.tex]: "List of operations **that ensure ordering and/or** delivery by Memory Ordering routines."? Maybe another phrasing for the caption?

**Review by BryantLam — CHANGES_REQUESTED** [review]
I have some questions or recommended edits. I'm generally happy about these changes though. :+1:

File: main_spec.tex [hunk 21]
```diff
@@ -473,7 +473,32 @@ \subsection{Memory Ordering Routines}\label{subsec:memory_order}
...
+      Blocking \OPR{Put}                 & X & X \\
+      Blocking \OPR{AMO}                 & X & X \\
+      Blocking \OPR{put-with-signal}     & X & X \\
+      Non-blocking \OPR{Put}             & X & X \\
+      Non-blocking \OPR{AMO}             & X & X \\
+      Non-blocking \OPR{put-with-signal} & X & X \\
+      Non-blocking \OPR{Get}             &   & X \\ \hline
+    \end{tabular}
+    \TableCaptionRef{List of operations ensured ordering and/or delivery by
+    Memory Ordering routines}
```
Comment (BryantLam) [inline main_spec.tex]:
```suggestion
    \TableCaptionRef{List of operations affected by ordering or delivery guarantees
    from Memory Ordering routines}
```

**Review by BryantLam — CHANGES_REQUESTED** [review]

**Review by BryantLam — APPROVED** [review]

File: main_spec.tex (hunk 21 above)
Comment (naveen-rn) [reply main_spec.tex]: Table modified - closing review as fixed

File: content/shmem_fence.tex (hunk 11 above)
Comment (naveen-rn) [reply content/shmem_fence.tex]: fixed.

File: content/shmem_quiet.tex (hunk 18 above)
Comment (naveen-rn) [reply content/shmem_quiet.tex]: used "outstanding operations" consistently

**Comment by naveen-rn** [comment]
@nspark can we close this PR. It looks like most requested changes are complete. We can add new fixes if needed as a separate PR? I'll wait till EoD and then try closing this before miidnight.

**Review by nspark — APPROVED** [review]
